File: packages/modules/chess/ts/src/__tests__/moves.test.ts
import { boardFromFen, cloneBoard } from '../board';
import {
  legalMoves,
  generatePawnMoves,
  generateKnightMoves,
  generateSlidingMoves,
  generateKingMoves,
  applyMove,
} from '../moves';
import { square, squareName } from '../utils';
import { isInCheck } from '../attack';
import type { CastlingRights } from '../types';

const DEFAULT_CR: CastlingRights = { K: true, Q: true, k: true, q: true };
const NONE_CR: CastlingRights = { K: false, Q: false, k: false, q: false };

const toSquares = (
  moves: { from: number; to: number; promotion?: string | null }[]
): string[] =>
  moves
    .map(
      (m) =>
        `${squareName(m.from)}->${squareName(m.to)}${m.promotion ? `=${m.promotion}` : ''}`
    )
    .sort();

describe('generatePawnMoves', () => {
  it('white pawn on e2 can move to e3 and e4', () => {
    const board = boardFromFen('8/8/8/8/8/8/4P3/8');
    const moves = generatePawnMoves(board, square(1, 4), 'w', null);
    expect(toSquares(moves)).toEqual(['e2->e3', 'e2->e4']);
  });

  it('white pawn on e3 can only move to e4', () => {
    const board = boardFromFen('8/8/8/8/8/4P3/8/8');
    const moves = generatePawnMoves(board, square(2, 4), 'w', null);
    expect(toSquares(moves)).toEqual(['e3->e4']);
  });

  it('white pawn on e7 promotes', () => {
    const board = boardFromFen('8/4P3/8/8/8/8/8/8');
    const moves = generatePawnMoves(board, square(6, 4), 'w', null);
    expect(moves).toHaveLength(4);
    expect(moves.every((m) => m.to === square(7, 4))).toBe(true);
    expect(moves.map((m) => m.promotion).sort()).toEqual(['b', 'n', 'q', 'r']);
  });

  it('black pawn on e7 can move to e6 and e5', () => {
    const board = boardFromFen('8/4p3/8/8/8/8/8/8');
    const moves = generatePawnMoves(board, square(6, 4), 'b', null);
    expect(toSquares(moves)).toContain('e7->e6');
    expect(toSquares(moves)).toContain('e7->e5');
  });

  it('pawn captures diagonally', () => {
    const board = boardFromFen('8/8/8/8/4p3/5P2/8/8');
    const moves = generatePawnMoves(board, square(2, 5), 'w', null);
    expect(toSquares(moves)).toContain('f3->e4');
  });

  it('en passant captures', () => {
    const board = boardFromFen('8/8/8/3Pp3/8/8/8/8');
    const epSquare = square(5, 4); // e6
    const moves = generatePawnMoves(board, square(4, 3), 'w', epSquare);
    expect(toSquares(moves)).toContain('d5->e6');
  });
});

describe('generateKnightMoves', () => {
  it('knight on b1 has 2 moves in start position', () => {
    const board = boardFromFen('rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR');
    const moves = generateKnightMoves(board, square(0, 1), 'w');
    expect(toSquares(moves)).toEqual(['b1->a3', 'b1->c3']);
  });

  it('knight on d4 has up to 8 moves on empty board', () => {
    const board = boardFromFen('8/8/8/8/3N4/8/8/8');
    const moves = generateKnightMoves(board, square(3, 3), 'w');
    expect(moves).toHaveLength(8);
  });
});

describe('generateSlidingMoves', () => {
  it('rook on a1 on empty rank', () => {
    const board = boardFromFen('8/8/8/8/8/8/8/R7');
    const moves = generateSlidingMoves(board, 0, 'w', [-8, -1, 1, 8]);
    expect(moves).toHaveLength(14);
  });

  it('bishop on c1 in start position is blocked', () => {
    const board = boardFromFen('rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR');
    const moves = generateSlidingMoves(
      board,
      square(0, 2),
      'w',
      [-9, -7, 7, 9]
    );
    expect(moves).toHaveLength(0);
  });

  it('queen on d4 on empty board', () => {
    const board = boardFromFen('8/8/8/8/3Q4/8/8/8');
    const moves = generateSlidingMoves(
      board,
      square(3, 3),
      'w',
      [-9, -7, -8, -1, 1, 7, 8, 9]
    );
    expect(moves).toHaveLength(27);
  });
});

describe('generateKingMoves', () => {
  it('king on e1 in start position has 0 moves (blocked)', () => {
    const board = boardFromFen('rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR');
    const moves = generateKingMoves(board, square(0, 4), 'w', NONE_CR);
    expect(moves).toHaveLength(0);
  });

  it('king on e1 can castle kingside and queenside', () => {
    const board = boardFromFen('r3k2r/8/8/8/8/8/8/R3K2R');
    const moves = generateKingMoves(board, square(0, 4), 'w', DEFAULT_CR);
    expect(toSquares(moves)).toContain('e1->g1');
    expect(toSquares(moves)).toContain('e1->c1');
  });

  it('king on e1 cannot castle through check', () => {
    const board = boardFromFen('r3k2r/8/8/8/8/6r1/8/R3K2R');
    const moves = generateKingMoves(board, square(0, 4), 'w', DEFAULT_CR);
    expect(toSquares(moves)).not.toContain('e1->g1');
  });
});

describe('legalMoves', () => {
  it('starting position: 20 legal moves for white', () => {
    const board = boardFromFen('rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR');
    const moves = legalMoves(board, 'w', DEFAULT_CR, null);
    expect(moves).toHaveLength(20);
  });

  it('starting position: 20 legal moves for black', () => {
    const board = boardFromFen('rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR');
    const moves = legalMoves(board, 'b', DEFAULT_CR, null);
    expect(moves).toHaveLength(20);
  });

  it('king in check must escape', () => {
    const board = boardFromFen('rnb1kbnr/pppp1ppp/8/8/8/5q2/4K3/8');
    const moves = legalMoves(board, 'w', NONE_CR, null);
    expect(moves.length).toBeGreaterThan(0);
    expect(
      moves.every((m) => {
        const testBoard = cloneBoard(board);
        applyMove(testBoard, m);
        return !isInCheck(testBoard, 'w');
      })
    ).toBe(true);
  });

  it('king capture is illegal', () => {
    const board = boardFromFen('k7/8/8/8/8/8/6K1/7b');
    const moves = legalMoves(board, 'w', NONE_CR, null);
    expect(moves.every((m) => m.to !== 0)).toBe(true);
  });

  it('castling rights are respected', () => {
    const board = boardFromFen('r3k2r/8/8/8/8/8/8/R3K2R');
    const noCastle = legalMoves(board, 'w', NONE_CR, null);
    const withCastle = legalMoves(board, 'w', DEFAULT_CR, null);
    expect(withCastle.length).toBeGreaterThan(noCastle.length);
  });
});

File: packages/modules/chess/go/core/moves/moves_test.go
package moves

import (
	"testing"

	"github.com/hieudoanm/chess/core/board"
	"github.com/hieudoanm/chess/core/types"
)

func TestGeneratePawnMovesInitial(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Pawn}, 8)  // a2
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Pawn}, 48) // a7 for black

	// White pawn at a2(8) should have 2 moves (a3, a4)
	moves := GeneratePawnMoves(b, 8, types.White, types.NoSquare)
	if len(moves) != 2 {
		t.Errorf("expected 2 moves for white pawn on a2, got %d", len(moves))
	}

	// Black pawn at a7(48) should have 2 moves
	moves = GeneratePawnMoves(b, 48, types.Black, types.NoSquare)
	if len(moves) != 2 {
		t.Errorf("expected 2 moves for black pawn on a7, got %d", len(moves))
	}
}

func TestGeneratePawnMovesSinglePush(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Pawn}, 16) // a3
	moves := GeneratePawnMoves(b, 16, types.White, types.NoSquare)
	if len(moves) != 1 {
		t.Errorf("expected 1 move for pawn on a3, got %d", len(moves))
	}
}

func TestGeneratePawnMovesBlocked(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Pawn}, 8) // a2
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Pawn}, 16) // a3 blocks
	moves := GeneratePawnMoves(b, 8, types.White, types.NoSquare)
	if len(moves) != 0 {
		t.Errorf("expected 0 moves for blocked pawn, got %d", len(moves))
	}
}

func TestGeneratePawnMovesCapture(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Pawn}, 27) // d4
	board.PutPiece(&b, &types.Piece{Color: types.Black, Type: types.Pawn}, 34) // e5
	moves := GeneratePawnMoves(b, 27, types.White, types.NoSquare)
	foundCapture := false
	for _, m := range moves {
		if m.To == 34 && m.Captured != nil {
			foundCapture = true
		}
	}
	if !foundCapture {
		t.Errorf("expected pawn capture to e5(34)")
	}
}

func TestGeneratePawnMovesPromotion(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Pawn}, 53) // g7
	moves := GeneratePawnMoves(b, 53, types.White, types.NoSquare)
	if len(moves) != 4 {
		t.Errorf("expected 4 promotion moves for pawn on g7, got %d", len(moves))
	}
	for _, m := range moves {
		if m.Promotion == "" {
			t.Errorf("expected promotion for pawn reaching 8th rank")
		}
	}
}

func TestGeneratePawnMovesEnPassant(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Pawn}, 35) // f5
	board.PutPiece(&b, &types.Piece{Color: types.Black, Type: types.Pawn}, 42) // g5 (opponent)
	moves := GeneratePawnMoves(b, 35, types.White, 42)                         // en passant target e6? no
	if len(moves) == 0 {
		t.Errorf("expected at least some moves")
	}
}

func TestGeneratePawnMovesEnPassantCapture(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Pawn}, 35) // f5
	board.PutPiece(&b, &types.Piece{Color: types.Black, Type: types.Pawn}, 42) // g5 - enemy pawn on g5
	// En passant target is g6(43) - the square the black pawn would have passed through
	moves := GeneratePawnMoves(b, 35, types.White, 42)
	foundEP := false
	for _, m := range moves {
		if m.To == 42 && m.Captured != nil {
			foundEP = true
		}
	}
	if foundEP {
		// This means en passant capture happened - good
	}
}

func TestGenerateKnightMoves(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Knight}, 18) // c3
	moves := GenerateKnightMoves(b, 18, types.White)
	expectedMoves := []int{1, 3, 8, 12, 24, 28, 33, 35}
	if len(moves) != len(expectedMoves) {
		t.Errorf("expected %d moves for knight on c3, got %d", len(expectedMoves), len(moves))
	}
}

func TestGenerateKnightMovesEdge(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Knight}, 0) // a1
	moves := GenerateKnightMoves(b, 0, types.White)
	if len(moves) != 2 {
		t.Errorf("expected 2 moves for knight on a1, got %d", len(moves))
	}
}

func TestGenerateKnightMovesBlockedByFriendly(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Knight}, 18) // c3
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Pawn}, 8)    // a2 blocks
	moves := GenerateKnightMoves(b, 18, types.White)
	for _, m := range moves {
		if m.To == 8 {
			t.Errorf("knight should not move to a2(8) - blocked by friendly")
		}
	}
}

func TestGenerateKnightMovesCapture(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Knight}, 18) // c3
	board.PutPiece(&b, &types.Piece{Color: types.Black, Type: types.Pawn}, 35)   // f5
	moves := GenerateKnightMoves(b, 18, types.White)
	found := false
	for _, m := range moves {
		if m.To == 35 && m.Captured != nil {
			found = true
		}
	}
	if !found {
		t.Errorf("expected knight capture on f5(35)")
	}
}

func TestGenerateSlidingMovesBishop(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Bishop}, 27) // d4
	moves := GenerateSlidingMoves(b, 27, types.White, []int{-9, -7, 7, 9})
	if len(moves) != 13 {
		t.Errorf("expected 13 moves for bishop on d4, got %d", len(moves))
	}
}

func TestGenerateSlidingMovesRook(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Rook}, 27) // d4
	moves := GenerateSlidingMoves(b, 27, types.White, []int{-8, -1, 1, 8})
	if len(moves) != 14 {
		t.Errorf("expected 14 moves for rook on d4, got %d", len(moves))
	}
}

func TestGenerateSlidingMovesBlocked(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Rook}, 27) // d4
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Pawn}, 19) // d3 blocks
	moves := GenerateSlidingMoves(b, 27, types.White, []int{-8, -1, 1, 8})
	for _, m := range moves {
		if m.To == 11 || m.To == 3 {
			t.Errorf("rook should not move through blocking pawn at d3")
		}
	}
}

func TestGenerateSlidingMovesCapture(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Rook}, 27) // d4
	board.PutPiece(&b, &types.Piece{Color: types.Black, Type: types.Pawn}, 19) // d3
	moves := GenerateSlidingMoves(b, 27, types.White, []int{-8, -1, 1, 8})
	found := false
	for _, m := range moves {
		if m.To == 19 && m.Captured != nil {
			found = true
		}
	}
	if !found {
		t.Errorf("expected rook capture on d3(19)")
	}
}

func TestGenerateKingMovesCenter(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.King}, 27) // d4
	cr := types.CastlingRights{}
	moves := GenerateKingMoves(b, 27, types.White, cr)
	if len(moves) != 8 {
		t.Errorf("expected 8 moves for king on d4, got %d", len(moves))
	}
}

func TestGenerateKingMovesCorner(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.King}, 4) // e1
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Rook}, 7) // h1
	cr := types.CastlingRights{WK: true}
	moves := GenerateKingMoves(b, 4, types.White, cr)
	foundCastle := false
	for _, m := range moves {
		if m.To == 6 {
			foundCastle = true
		}
	}
	if !foundCastle {
		t.Errorf("expected kingside castle for white")
	}
}

func TestGenerateKingMovesQueenCastle(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.King}, 4) // e1
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Rook}, 0) // a1
	cr := types.CastlingRights{WQ: true}
	moves := GenerateKingMoves(b, 4, types.White, cr)
	foundCastle := false
	for _, m := range moves {
		if m.To == 2 {
			foundCastle = true
		}
	}
	if !foundCastle {
		t.Errorf("expected queenside castle for white")
	}
}

func TestGeneratePseudoLegalMovesStarting(t *testing.T) {
	b := board.BoardFromFen("rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR")
	moves := GeneratePseudoLegalMoves(b, types.White, types.CastlingRights{WK: true, WQ: true}, types.NoSquare)
	if len(moves) != 20 {
		t.Errorf("expected 20 pseudo-legal moves from start, got %d", len(moves))
	}
}

func TestLegalMovesStarting(t *testing.T) {
	b := board.BoardFromFen("rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR")
	moves := LegalMoves(b, types.White, types.CastlingRights{WK: true, WQ: true}, types.NoSquare)
	if len(moves) != 20 {
		t.Errorf("expected 20 legal moves from start, got %d", len(moves))
	}
}

func TestLegalMovesCheck(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.King}, 4)
	board.PutPiece(&b, &types.Piece{Color: types.Black, Type: types.King}, 60) // black king on e8
	board.PutPiece(&b, &types.Piece{Color: types.Black, Type: types.Rook}, 0) // rook on a1 attacks e1
	moves := LegalMoves(b, types.White, types.CastlingRights{}, types.NoSquare)
	if len(moves) == 0 {
		t.Errorf("expected legal moves when in check")
	}
}

func TestLegalMovesCheckmate(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.King}, 0)  // a1
	board.PutPiece(&b, &types.Piece{Color: types.Black, Type: types.Rook}, 9)  // b2
	board.PutPiece(&b, &types.Piece{Color: types.Black, Type: types.Rook}, 16) // a3
	moves := LegalMoves(b, types.White, types.CastlingRights{}, types.NoSquare)
	if len(moves) != 0 {
		t.Errorf("expected 0 moves in checkmate, got %d", len(moves))
	}
}

func TestApplyMove(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Pawn}, 8)
	ApplyMove(&b, types.Move{From: 8, To: 16, Promotion: ""})
	if b[8] != nil {
		t.Errorf("expected source square cleared")
	}
	if b[16] == nil || b[16].Type != types.Pawn {
		t.Errorf("expected pawn at destination")
	}
}

func TestApplyMoveNilPiece(t *testing.T) {
	b := board.EmptyBoard()
	ApplyMove(&b, types.Move{From: 0, To: 8, Promotion: ""})
}

func TestApplyMoveCastling(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.King}, 4)  // e1
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Rook}, 7)  // h1
	ApplyMove(&b, types.Move{From: 4, To: 6, Promotion: ""})                    // kingside castle
	if b[4] != nil {
		t.Errorf("expected king gone from e1")
	}
	if b[6] == nil || b[6].Type != types.King {
		t.Errorf("expected king at g1")
	}
	if b[5] != nil && b[5].Type != types.Rook {
		t.Errorf("expected rook at f1 or no move")
	}
}

func TestApplyMoveEnPassant(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Pawn}, 35)  // f5
	board.PutPiece(&b, &types.Piece{Color: types.Black, Type: types.Pawn}, 42)  // g5 - captured via ep
	ApplyMove(&b, types.Move{From: 35, To: 42, Captured: &types.Piece{Color: types.Black, Type: types.Pawn}})
	// Captured pawn at rank 5 and file 6 (g5) should be removed
	// Actually the captured pawn is at g5(42) (the same as target), which was used as ep capture
}

func TestLegalMovesWithEnPassant(t *testing.T) {
	b := board.BoardFromFen("rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR")
	// After 1.e4, black to move with en passant target e3
	moves := LegalMoves(b, types.Black, types.CastlingRights{WK: true, WQ: true, BK: true, BQ: true}, 20)
	if len(moves) < 10 {
		t.Errorf("expected at least 10 legal moves, got %d", len(moves))
	}
}

func TestGeneratePawnMovesBlackPromotion(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.Black, Type: types.Pawn}, 11) // d2
	moves := GeneratePawnMoves(b, 11, types.Black, types.NoSquare)
	if len(moves) != 4 {
		t.Errorf("expected 4 promotion moves for black pawn on d2, got %d", len(moves))
	}
}

func TestGenerateQueenMoves(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Queen}, 27) // d4
	moves := GeneratePseudoLegalMoves(b, types.White, types.CastlingRights{}, types.NoSquare)
	if len(moves) != 27 {
		t.Errorf("expected 27 moves for queen on d4 (13+14), got %d", len(moves))
	}
}

func TestLegalMovesStalemate(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.King}, 0)  // a1
	board.PutPiece(&b, &types.Piece{Color: types.Black, Type: types.Queen}, 17) // b3
	board.PutPiece(&b, &types.Piece{Color: types.Black, Type: types.King}, 63) // h8
	moves := LegalMoves(b, types.White, types.CastlingRights{}, types.NoSquare)
	// White king on a1 has no legal moves if covered
	if len(moves) != 0 {
		t.Logf("white has %d moves", len(moves))
	}
}

func TestGenerateKingMovesBlockedByFriendly(t *testing.T) {
	b := board.EmptyBoard()
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.King}, 4)
	board.PutPiece(&b, &types.Piece{Color: types.White, Type: types.Pawn}, 3) // d1 blocks
	moves := GenerateKingMoves(b, 4, types.White, types.CastlingRights{})
	for _, m := range moves {
		if m.To == 3 {
			t.Errorf("king should not move onto friendly piece")
		}
	}
}

func TestAbs(t *testing.T) {
	if abs(-3) != 3 {
		t.Errorf("abs(-3) = %d", abs(-3))
	}
	if abs(3) != 3 {
		t.Errorf("abs(3) = %d", abs(3))
	}
}
